providers/defltprov.c

```diff
@@ -477,7 +477,11 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
 # endif
     { PROV_NAMES_EC, "provider=default", ossl_ec_asym_kem_functions },
 #endif
-    { PROV_NAMES_MLKEM768, "provider=default", ossl_mlkem768_asym_kem_functions },
+#ifndef OPENSSL_NO_ML_KEM
+    { PROV_NAMES_ML_KEM_512, "provider=default", ossl_ml_kem_asym_kem_functions },
+    { PROV_NAMES_ML_KEM_768, "provider=default", ossl_ml_kem_asym_kem_functions },
+    { PROV_NAMES_ML_KEM_1024, "provider=default", ossl_ml_kem_asym_kem_functions },
```

Comment:
Hmm... we could actually have only a ML-KEM kem provider and use the query_operation_name() on the keymgmt to return ML-KEM. This would be IMO the right thing. However I propose to postpone this change to another issue after this PR is merged as is.

Reply:
I still don't see a problem. The encap/decap provider does not create keys, it just uses the keys it is given to perform encapsulation/decapsulation (with an optional seed).

Nothing about the code needs to know or care about which variant the key is for, the key already has that data. Why require separate method blocks here, just to check that the key type matches the provider instance?

Suppose someone manages to use the EVP_KEM object from ML-KEM-x with a key for ML-KEM-y, is that actually a problem we need to be concerned about?

If they get the right answer, then clearly they wanted the operation corresponding to the key, and the reuse of an ML-KEM object from a related KEM fetch is harmless.

If they get the wrong answer, that's unsupported, they were supposed to pass a key of the right type.

In any case, the EVP layer is responsible for checking that the EVP_PKEY_CTX is compatible with the EVP_KEM, along the lines of: https://github.com/openssl/openssl/blob/e0ea913f11cf64d000556bbf7cb9f8acdf6be4cb/crypto/evp/signature.c#L548-L586
Where, absent a `query_key_types` method in the signature, the keymgmt pointers are compared.

If EVP_KEMs are going to become objects usable in their own right, presumably they'll need new methods to return the supported key types, and it is at that point that it'll be clear that a variant-specific function is needed (or perhaps still not, if we really choose to not care, which ML-KEM you fetch, and document that it is the key that determines what happens).

Which is not to say, that I'm refusing to duplicate the function block three ways, but that assumes that something will be comparing these pointers to determine compatibility, which is not the approach taken with signatures, instead a specific method is called to return the compatible key types.

So I think it is too early to say whether this is warranted, and it seems plausible that appropriate changes would be needed anyway *when* that becomes necessary.

Reply:
All that said, I did notice the the KEM operation context structure carries a libctx value that is saved and never used. After the PR is merged, I'm inclined to drop it, unless you strongly feel it should be done now.

The relevant libctx is in the key (saved during key creation), and at that time used to locate SHA3 methods and for the RNG used to generate the key material. Later the same libctx is used in random number generation.  I don't think it makes sense for KEM ops to override the RNG used when creating the key just for the duration of encap/decap.

If, however, it is in fact expect that the libctx of the key **should** be preëmpted by the libctx of the encap/decap caller, (with its sole use being random number generation), then we'd need to pass the libctx into the internal encap/decap methods, rather than use the saved libctx from the key (which wold then not need to be saved).

So let me know if unexpectedly the KEM CTX should determine the RNG separately from key management.

Reply:
This is actually a good philosophical question - should EVP_PKEYs be usable across different libctxes? When they contain just pure data or provider references which are associated with a concrete key management pointer, there is no issue. However now the libctx is internal to the provider-side key data. We AFAIK do not have this situation with other key types.

I would not try to solve this problem in this PR though. We should IMO merge this PR as is and have a subsequent issue to think about this.

Reply:
So "fix it now" or later? There isn't yet any way to do what you're concerned about, and I still think that the EVP layer will have to know whether the PKEY and KEM objects are compatible using some new feature of KEM objects, so the issue can probably wait till then.

Reply:
> A possible answer, as suggest by @t8m, is to only have the variants in the key manager and not in the kem at all. So in the kem we just fetch "ML-KEM" which works for all types of keys. Unless we think that is too confusing - in which case I think we should have different ML-KEM instances for each key type that validate we have the correct key type when they are used.

Let's not recreate the pain the `EC` mistake has caused.  We should be fetching different things.
That they have the same underlying implementation is fine.

Reply:
Agree with @paulidale - they are different. Fetching "ML-KEM" would make no sense IMHO here. We treat all three feature sets independently. The implementation details are just implementation details - they should not impact the users use as independent algorithms.

Reply:
Sure, so no consolidation into a single logical algorithm, but I still see no immediate reason to have separate function blocks now. Each algorithm has its own name, and if EVP wants some day to compare they key with the provider for compatibility, a new entry point willl be needed at **that** time. At the moment, I can't think of any reason to duplicate the array.

Reply:
Bottom line Matt, thanks for raising this point, it makes sense to keep this in mind for the future, but for now I think the PR is fine, will deal with this if and when stand-alone KEM objects are a thing.
I think I've addressed all your comments, please let me know if there's anything else.

Reply:
@paulidale I agree, but there is no code that does not care which is used at this time, because there is no way to actually use any fetched KEM. When we add support to use fetched KEMs, we'll have to add a check that, for example, nobody is passing an **ECX** key to the **ML-KEM** object, such checks will require new methods, and these will trigger the need to possibly separate the tables.